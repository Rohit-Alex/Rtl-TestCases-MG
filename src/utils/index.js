export const STATUS = {
    transactionStatus: true
}
export const constructMessage = (logLevel, message) => {
    return {
        data: {
            logLevel,
            message
        }
    }
}

export const fetchMethod = (url) => fetch(url).then(res => {
    return new Promise((resolve, reject) => {
        if (res.status === 200 && res.ok) {
            resolve(res.json())
        }
        else reject('Api failure')
    })
})