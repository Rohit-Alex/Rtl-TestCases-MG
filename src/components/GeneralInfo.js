import React, { useState, useEffect } from 'react'
import { Card } from 'antd'
import DescriptionWrapper from './Descriptions'
import SpinnerContainer from './Spinner'
import { fetchMethod } from '../utils'

const GeneralInfo = () => {
    const [rowData, setRowData] = useState([])
    const [isLoading, setIsLoading] = useState(true)

    const getAllData = async () => {
        try {
            const data = await fetchMethod('https://jsonplaceholder.typicode.com/todos')
            setRowData(data.generalInfo)
        }
        catch(err) {
            console.log(err, 'err')
        }
        finally {
            setIsLoading(false)
        }
    }

    useEffect(() => {
        getAllData()
    }, [])


    return (
        <div className="generalInfo" data-testid='generalInfo-testid'>
            {isLoading ? (
                <SpinnerContainer isLoading={isLoading} />
            ) : (
                <Card bordered={false} className="audit-hoverable" title="General Information">
                    <DescriptionWrapper
                        options={{ ...rowData, transactionStatus: rowData.transactionStatus ? ('CREATED') : ('NOT_CREATED') }}
                        column={6}
                        layout="vertical"
                        bordered
                    />

                </Card>
            )}
        </div>
    )
}

export default GeneralInfo