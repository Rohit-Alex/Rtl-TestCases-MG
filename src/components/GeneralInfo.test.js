/* eslint-disable testing-library/no-render-in-setup */
/* eslint-disable testing-library/no-debugging-utils */
/* eslint-disable testing-library/no-unnecessary-act */
import React from 'react'

import GeneralInfo from './GeneralInfo'
import { cleanup, render, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react'
import '@testing-library/jest-dom'
import { fetchMethod } from '../utils'
import { act } from 'react-dom/test-utils'


jest.mock('../utils')

// jest.mock('../utils', () => ({
//     ...jest.requireActual('../utils'),
//     fetchMethod: () => Promise.resolve({
//         // json: () => Promise.resolve({
//         //     generalInfo: {
//         //         options: [{ label: 'label1', value: 'customValue1' }],
//         //         transactionStatus: true
//         //     }
//         // })
//         generalInfo: {
//             options: [{ label: 'label1', value: 'customValue1' }],
//             transactionStatus: true
//         }
//     })
// }));


window.matchMedia =
    window.matchMedia ||
    (() => {
        return {
            matches: false,
            addListener() { },
            removeListener() { },
        }
    })

describe("Test GeneralInfo component", () => {


    afterEach(() => {
        cleanup()
        fetchMethod.mockReset()
    })


    it('should show loader till data is fetched', async () => {
        act(() => {
            render(<GeneralInfo />)
        })
        expect(screen.getByTestId('generalInfo-testid')).toBeInTheDocument()
        expect(screen.getByTestId('spinner-container')).toBeInTheDocument()
        await waitForElementToBeRemoved(() => screen.queryByTestId('spinner-container'))
    })

    it('should show data(CREATED) when data is fetched created', async () => {
        await act(async () => {
            fetchMethod.mockReturnValue({
                generalInfo: {
                    "formula": '{"formula":{"expressionSet":{"operator":"*","operands":[{"key":"$/data/sellerOrder/totals[?(@/type=="SHIPPING_TOTAL")]/amount","valueType":"Number","specificType":"Money","attributeType":"SYSTEM"}]}}}',
                    transactionStatus: true
                }
            })
            render(<GeneralInfo />)
        })
        expect(screen.getByTestId('generalInfo-testid')).toBeInTheDocument()
        expect(screen.getByText('formula')).toBeInTheDocument()
        expect(screen.getByTestId('viewmodal-test-id')).toBeInTheDocument()
        expect(screen.getByText('General Information')).toBeInTheDocument()
        expect(screen.getByRole("table")).toBeInTheDocument()
        expect(screen.getByTestId('badge-testid').textContent).toBe('CREATED')
        // await waitFor(() => {
        //     expect(screen.getByRole("table")).toBeInTheDocument()
        //     expect(screen.getByText('CREATED')).toBeInTheDocument()
        // })
    })

    it('should show data(NOT_CREATED) when data is fetched not created', async () => {
        await act(async () => {
            fetchMethod.mockReturnValue({
                generalInfo: {
                    customKey: 'customValue',
                    transactionStatus: false
                }
            })
            render(<GeneralInfo />)
        })
        expect(screen.getByTestId('generalInfo-testid')).toBeInTheDocument()
        expect(screen.getByText('General Information')).toBeInTheDocument()
        expect(screen.getByRole("table")).toBeInTheDocument()
        expect(screen.getByText('customKey')).toBeInTheDocument()
        expect(screen.getByText('customValue')).toBeInTheDocument()
        expect(screen.getByTestId('badge-testid').textContent).toBe('NOT_CREATED')
        // await waitFor(() => {
        //     const dataContainer = screen.getByRole("table")
        //     expect(screen.getByText('NOT_CREATED')).toBeInTheDocument()
        //     expect(dataContainer).toBeInTheDocument()
        // })
    })
})
