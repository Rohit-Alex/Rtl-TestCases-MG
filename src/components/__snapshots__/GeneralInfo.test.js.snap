// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GeneralInfo component should load component 1`] = `
<div
  className="generalInfo"
  data-testid="generalInfo-testid"
>
  <div
    className="ant-spin ant-spin-spinning"
    data-testid="spinner-container"
  >
    <span
      className="ant-spin-dot ant-spin-dot-spin"
    >
      <i
        className="ant-spin-dot-item"
      />
      <i
        className="ant-spin-dot-item"
      />
      <i
        className="ant-spin-dot-item"
      />
      <i
        className="ant-spin-dot-item"
      />
    </span>
  </div>
</div>
`;
